sql/expression/arithmetic.go

```diff
@@ -0,0 +1,411 @@
+func NewIntDiv(left, right sql.Expression) *Arithmetic {
+	return NewArithmetic(left, right, sqlparser.IntDivStr)
+}
+
+// NewMod creates a new Arithmetic % sql.Expression.
+func NewMod(left, right sql.Expression) *Arithmetic {
+	return NewArithmetic(left, right, sqlparser.ModStr)
+}
+
+func (a *Arithmetic) String() string {
+	return fmt.Sprintf("%s %s %s", a.Left, a.op, a.Right)
+}
+
+// Type returns the greatest type for given operation.
+func (a *Arithmetic) Type() sql.Type {
+	switch a.op {
+	case sqlparser.PlusStr, sqlparser.MinusStr, sqlparser.MultStr, sqlparser.DivStr:
+		if sql.IsInteger(a.Left.Type()) && sql.IsInteger(a.Right.Type()) {
+			if sql.IsUnsigned(a.Left.Type()) && sql.IsUnsigned(a.Right.Type()) {
+				return sql.Uint64
+			}
+			return sql.Int64
+		}
+		return sql.Float64
+
+	case sqlparser.ShiftLeftStr, sqlparser.ShiftRightStr:
+		return sql.Uint64
+
+	case sqlparser.BitAndStr, sqlparser.BitOrStr, sqlparser.BitXorStr, sqlparser.IntDivStr, sqlparser.ModStr:
+		if sql.IsUnsigned(a.Left.Type()) && sql.IsUnsigned(a.Right.Type()) {
+			return sql.Uint64
+		}
+		return sql.Int64
+	}
+
+	return sql.Float64
+}
+
+// TransformUp implements the Expression interface.
+func (a *Arithmetic) TransformUp(f sql.TransformExprFunc) (sql.Expression, error) {
+	l, err := a.Left.TransformUp(f)
+	if err != nil {
+		return nil, err
+	}
+
+	r, err := a.Right.TransformUp(f)
+	if err != nil {
+		return nil, err
+	}
+
+	return f(NewArithmetic(l, r, a.op))
+}
+
+// Eval implements the Expression interface.
+func (a *Arithmetic) Eval(ctx *sql.Context, row sql.Row) (interface{}, error) {
+	span, ctx := ctx.Span("expression.(" + a.op + ")")
+	defer span.Finish()
+
+	lval, rval, err := a.eval2(ctx, row)
+	if err != nil {
+		return nil, err
+	}
+
+	lval, rval, err = a.convert2(lval, rval)
+	if err != nil {
+		return nil, err
+	}
+
+	switch a.op {
+	case sqlparser.PlusStr:
+		return plus(lval, rval)
+	case sqlparser.MinusStr:
+		return minus(lval, rval)
+	case sqlparser.MultStr:
+		return mult(lval, rval)
+	case sqlparser.DivStr:
+		return div(lval, rval)
+	case sqlparser.BitAndStr:
+		return bitAnd(lval, rval)
+	case sqlparser.BitOrStr:
+		return bitOr(lval, rval)
+	case sqlparser.BitXorStr:
+		return bitXor(lval, rval)
+	case sqlparser.ShiftLeftStr:
+		return shiftLeft(lval, rval)
+	case sqlparser.ShiftRightStr:
+		return shiftRight(lval, rval)
+	case sqlparser.IntDivStr:
+		return intDiv(lval, rval)
+	case sqlparser.ModStr:
+		return mod(lval, rval)
+	}
+
+	return nil, errUnableToEval.New(lval, a.op, rval)
+}
+
+func (a *Arithmetic) eval2(ctx *sql.Context, row sql.Row) (interface{}, interface{}, error) {
+	lval, err := a.Left.Eval(ctx, row)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	rval, err := a.Right.Eval(ctx, row)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return lval, rval, nil
+}
+
+func (a *Arithmetic) convert2(lval interface{}, rval interface{}) (interface{}, interface{}, error) {
```

Comment:
why `convert2`?

Reply:
the same reason like eval2. It's a internal stuff working for both (left and right)

Reply:
it's like a notation `eval/2` (known from many langs.)

Reply:
just to make it less confusing I renamed these function to `evalLeftRight` and `convertLeftRight`